Implementing System.py's Registry() (WIP, Some Questions)

HayashiEsme opened the pull request:
I hope I interpreted the code correctly. This was what I thought the code should do more or less. Please correct me if I've got it wrong. 

I've got questions about individual bits in the implementation so I guess I'll leave it in comments

HayashiEsme commented on wslpy/system.py:
```diff
@@ -63,7 +63,7 @@ def shellEnvVar(input):
     ----------
     input : str
-        string of a shell environment variable key.
+        string of a shell environment variable's location.
```
Ignore this, this is a typo and will be revered to original in the next commit

HayashiEsme commented on wslpy/system.py:
```diff
@@ -78,4 +78,67 @@ def shellEnvVar(input):
+    routput, err = p.communicate()
+    #DUMMY CODE ENDS 
+
+    # Given a valid registry path, retrieves the value of an entry in the registry.
```
Are there plans to maintain the function's comments with the same format as in the triple quotes?

HayashiEsme replied:
See #15 on discussion with respect to comment format implementation

HayashiEsme commented on wslpy/system.py:
```diff
@@ -78,4 +78,67 @@ def shellEnvVar(input):
+        if query != "value.": #Error after regex #TODO: Check with Patrick on how he wants to handle this
+            return query #Expected one
+        else:
+            raise KeyError("The following error propogated from Registry:" + routput.decode("utf-8").rstrip()) #TODO: 
```
(This is where dummy code comes in) 

I was thinking about how to handle invalid entries, and I wasn't sure how I would go about implementing this considering that the result from __regInfoFetch()__ is pre-formatted. Are there plans for this to become a feature?/Should I try to implement this?

patrick330602 replied:
Notice the command from 83-85:

```
cmd = u"reg.exe query \""+input+"\" /v \""+key+u"\" 2>&1"
p = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE)
routput, err = p.communicate()
```

`p.communicate` will return both the output and error. That might be helpful for you.

HayashiEsme replied:
I had a thought, I'm thinking of amending __regInfoFetch()__ to return either the error or the value like so:

```
def __regInfoFetch__(input, key):
    cmd = u"reg.exe query \""+input+"\" /v \""+key+u"\" 2>&1"
    p = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE)
    routput, err = p.communicate()
    output = (routput.decode("utf-8").rstrip().split())[-1]

    if err == None: #No error
        return output
    else:
        return err
```

Actually: would you rather me make the judgement on design decisions like these?

HayashiEsme replied:
Thanks for this, I must've glossed over this for some reason. 

Unfortunately err doesn't seem to bring out errors and I had to implement a workaround (see https://github.com/wslutilities/wslpy/commit/6fc47610d174f222973387da362e454b654ac5f0 ) to handle this. Perhaps this is an issue with upstream Python on handling errors from Powershell via WSL

HayashiEsme commented on wslpy/system.py:
```diff
@@ -78,4 +78,67 @@ def shellEnvVar(input):
+    except KeyError as err:
+        print(err)
+ 
+#### Test Cases
```
These are for my own reference, I'll delete it when I have a final version ready to merge

patrick330602 replied:
test cases should be placed under test in the future, currently, there is no test but I will hook up test suite to help you make it work -- once you are done

HayashiEsme commented on wslpy/system.py:
```diff
@@ -78,4 +78,67 @@ def shellEnvVar(input):
+    #DUMMY CODE ENDS 
+
+    # Given a valid registry path, retrieves the value of an entry in the registry.
+    # Eg: registry("HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\Session Manager\Environment","OS") returns "WINDOWS_NT"
```
Did I understand the code right? Is this what the library is designed to take in?

patrick330602 replied:
Yep, this is right. Also if possible, the optimal goal is to return its type too:
![image](https://user-images.githubusercontent.com/15316889/89114583-aa04df00-d4b0-11ea-97e6-a2985a9895c2.png)

HayashiEsme replied:
Ah, shall we return this as an array or a class object?

HayashiEsme replied:
Added in the new commit, see https://github.com/wslutilities/wslpy/pull/13/commits/6fc47610d174f222973387da362e454b654ac5f0

shubhsboson commented:
Sorry for dropping in , just wanted to mention that I had no idea @HayashiEsme was working on registry(input,key) , and I already implemented it , so maybe we can work together or communicate properly so things like this don't happen again😅😅

HayashiEsme commented:
@shubhsboson You should've noticed that I had made everyone aware of what I was working on a week before you started work, thus communicating properly. https://github.com/wslutilities/wslpy/issues/7#issuecomment-662851316 Please check on who's working on this next time. I found it kind of rude for you to accuse me of not "communicating properly"

shubhsboson commented:
Hey! I am really sorry for being rude at any point of time , I'll leave this here to not create any more problems.

HayashiEsme commented:
Thanks that has been greatly appreciated :+1:

HayashiEsme commented:
I've made changes to `__regInfoFetch__()` and `registry()` that addresses the issues raised here in the pull request. I think that working on registry() has shown that discussion with relation to implementation and doc formatting will carry onto `__shellEnvVarList()__` and `shellEnvVar()` ( [which I've now claimed](https://github.com/wslutilities/wslpy/issues/7)) so I'd like to continue working on those using the same pull request if that hasn't reduced clarity in commit history. Given that I'm waiting to implement a test suite too, I've left my test cases in there for when that happens. Shall we close this issue and merge when all of this discussed is done? :)

patrick330602 commented:
Hi,

Sorry to be late, I will try my best to spare some time to review the code tonight.

HayashiEsme commented:
I noticed there's a merge conflict arising from the issues discussed. Shall we decide on what how to proceed when you've got a little more time? :)

patrick330602 commented:
> I noticed there's a merge conflict arising from the issues discussed. Shall we decide on what how to proceed when you've got a little more time? :)

Don't need to, I am doing it on cli right now
